core/modules/views/handlers/views_handler_field_bulk_form.inc

```diff
@@ -98,13 +98,19 @@ class views_handler_field_bulk_form extends views_handler_field {
     // Render checkboxes for all rows.
     $form[$this->options['id']]['#tree'] = TRUE;
     foreach ($this->view->result as $row_index => $row) {
+      // Store the row index to retrieve the entity from the results.
+      $this->view->row_index = $row_index;
```

Comment:
Maybe a silly question, but why is this necessary?

Reply:
This is needed because `views_handler_field_entity:get_value()` needs to know the row index in order to retrieve the value of the ID field for the entity. That row index points at the correct row in the results of the query. `get_value()` is invoked in `views_handler_bulk_form:views_form()` in order to build the checkboxes with the correct ID of the real field.

Reply:
A bit more explanation - there is no assurance that the $values variable actually contains the value of $field when invoking `views_handler_field_entity:get_value()` - because the value of `real_field` may or may not have been included in the query. Therefore get_value needs to retrieve the actual entity (which was cache during  `views_handler_field_entity:pre_render()` in order to obtain the value of that field.

Reply:
I see (sort of) :wink:

What baffled me was that inside a foreach the same property gets overridden again and again.

I'll probably have to locally stuff the code with debug()s - hopefully then I'll understand.